### ISSUE TITLE
@PolyNull in a functional-interface type argument resolves to @NonNull for a lambda's return

### ISSUE THREAD

**Issue by wmdietl** [eisop/checker-framework#2030]
`@PolyNull` in a functional-interface type argument is resolved to `@NonNull` for a lambda's return, so returning `null` from the lambda is a false positive.

The most visible consequence is that the standard `Map` idioms are rejected:

```java
import java.util.Map;

class Q {
  void m(Map<String, String> map) {
    map.computeIfAbsent("k", k -> null);
  }
}
```

```
Q.java:4: error: [return.type.incompatible] incompatible types in return.
    map.computeIfAbsent("k", k -> null);
                                  ^
  type of expression: @Nullable NullType
  method return type: @NonNull String
```

The annotated JDK declares

```java
default @PolyNull V computeIfAbsent(K key, Function<? super K, ? extends @PolyNull V> mappingFunction)
```

which is the intended specification: a mapping function that may return `null` yields a method result that may be `null`. But the lambda's return type is checked against `@NonNull V`, not against the poly-resolved type, so the `null` return is reported.

### Minimal reproduction, independent of the JDK

```java
import java.util.function.BiFunction;

import org.checkerframework.checker.nullness.qual.Nullable;
import org.checkerframework.checker.nullness.qual.PolyNull;

class PolyLambda<K, V> {
  @PolyNull V poly(K k, BiFunction<? super K, ? super @Nullable V, ? extends @PolyNull V> f) {
    throw new Error();
  }

  @Nullable V nullable(K k, BiFunction<? super K, ? super @Nullable V, ? extends @Nullable V> f) {
    throw new Error();
  }

  static void use(PolyLambda<String, String> m) {
    m.poly("k", (k, v) -> null);     // error: method return type: @NonNull String
    m.nullable("k", (k, v) -> null); // no error
  }
}
```

The two declarations differ only in `@PolyNull` versus `@Nullable` on the functional interface's return type argument. The `@Nullable` one behaves as expected, so the problem is in how the polymorphic qualifier is instantiated for a lambda body, not in checking the lambda itself.

### Scope

`? extends @PolyNull V` appears 49 times across nine files of the annotated JDK -- `Map`, `HashMap`, `Hashtable`, `TreeMap`, `Collections`, `ImmutableCollections`, `ConcurrentMap`, `ConcurrentHashMap`, `ConcurrentSkipListMap` -- covering `compute`, `computeIfAbsent`, `computeIfPresent`, and `merge`. Returning `null` from those lambdas is how a caller asks for the entry to be removed (or not created), so this affects ordinary code.

`Map.compute` and `Map.computeIfPresent` happen to be declared with plain `@Nullable` rather than `@PolyNull` (see the `@CFComment` at `Map.java:719`), so they do *not* reproduce it; `Map.computeIfAbsent` and every `HashMap`/`ConcurrentHashMap` override do.

### Found by

A Checker-Framework-versus-NullAway comparison on Caffeine. Of the 908 positions the Nullness Checker reported and NullAway did not, this pattern accounts for about 30 findings -- the largest group among those that are genuine dataflow findings rather than declaration-level generics mismatches.

**Comment by wmdietl** [comment]
Diagnosis, plus two attempted fixes that do not work. Recording both so the next attempt starts further along.

### Where the `@NonNull` comes from

`AbstractQualifierPolymorphism.resolve(MethodInvocationTree, ...)` collects an instantiation by matching argument types against formal parameter types:

```java
List<AnnotatedTypeMirror> arguments =
        CollectionsPlume.mapList(atypeFactory::getAnnotatedType, tree.getArguments());
AnnotationMirrorMap<AnnotationMirror> instantiationMapping = collector.visit(arguments, parameters);
```

For a lambda argument, `getAnnotatedType` returns the lambda's functional interface type, which `getFunctionalInterfaceType` derives from the invocation's own formal parameter -- the very type being resolved. So the qualifier is instantiated from a type that defaulting produced, and that instantiation then becomes the target the lambda body is checked against.

When no instantiation is found at all, the two fallbacks disagree:

- `replacer` (`DefaultQualifierPolymorphism.replace` with an empty map) uses the **top** qualifier, "since there is no further information to deduce".
- `completer` uses the **bottom** qualifier and, on a type variable or wildcard, removes the polymorphic qualifier *without putting anything in its place*, leaving the type to defaulting.

`resolve` picks the first only for a varargs method invoked with zero varargs actuals, and the second otherwise. `? extends @PolyNull V` takes the second path: `@PolyNull` is dropped from a type-variable use, `V` substitutes to `@NonNull String`, and the lambda's `return null` is rejected.

### This is the same bug as #1633

`checker/tests/nullness/java8/Issue1633.java:42-52` already documents it:

```java
void foo41(Optional1633<String> o) {
    // This is a false postive because inference doesn't work with poly qualifiers.
    // :: error: (return.type.incompatible)
    @Nullable String str3 = o.orElseGetPolyNull(() -> null);
}
```

Any fix should make those two expected errors go away.

### Attempt 1: treat a lambda-only invocation like the zero-varargs case

Exclude lambda and method-reference arguments from the collection (they are "not pertinent to applicability" in the language's own sense), and when that leaves nothing, route the empty mapping to `replacer` (top) rather than `completer`.

Fixes this issue and both `Issue1633` cases. **Regresses** `checker/tests/nullness/Issue1817.java:42`: `map.computeIfAbsent(k, __ -> new ArrayList<Object>())` now returns `@Nullable`, so dereferencing the result is an error. Top is too blunt -- it discards the information that the lambda plainly returns non-null.

### Attempt 2: instantiate from the lambda's body

Keep excluding the lambda as a whole, but match the type of an expression-bodied lambda's body against the polymorphic return type of the function type its formal parameter describes. This is the one place the information can soundly come from, since the body's type is written by the programmer rather than derived from the target.

Handles all three nullness cases correctly -- `__ -> new ArrayList<>()` gives `@NonNull`, `k -> null` and `() -> null` give `@Nullable`. But it **regresses other checkers** that put polymorphic qualifiers on functional interfaces, in `:checker:test`:

- `guieffect/Java8Lambdas.java` -- ten `call.invalid.ui`, `assignment.type.incompatible` and `return.type.incompatible` diagnostics
- `nonempty/Streams.java` -- four `assignment.type.incompatible`
- `all-systems` and `all-systems/java8inference` under `OptionalTest`

So matching only the *return* position of the function type is not enough: a lambda's parameter types participate in these hierarchies too, and pairing body against return alone changes the instantiation those checkers rely on.

### What this suggests

The fix probably has to instantiate the polymorphic qualifier from the whole lambda -- parameters from the target as today, return from the body -- rather than from either one alone, and to leave the `completer`/`replacer` split alone. Both attempts above are reverted; nothing is committed.

Related: the `@PolyNull` declarations this affects are `? extends @PolyNull V` on `compute`, `computeIfAbsent`, `computeIfPresent` and `merge` -- 49 occurrences across nine files of the annotated JDK. `Map.compute` and `Map.computeIfPresent` use plain `@Nullable` instead (see the `@CFComment` at `Map.java:719`) and so are unaffected.

**Comment by cpovirk** [comment]
(You may also have seen typetools 7340, a response to a typetools change to use `@Nullable` on the `Map` methods.)
